Gatekeeper implementation with 'standalone bug' fixed

ndev2 opened the pull request.

milesdevis committed:
net/sched: add gkprio scheduler//Checkpatch.pl checks

cjdoucette commented:
I know this is not the most up-to-date version of your code, but there is one large fix that I think you should apply before rebasing here.

After looking through the code, I see that each pfifo is really just using the linked list functionality that's built into each `struct sk_buff`. Since each pfifo is just a linked list, changing the size of the list (via `fifo_set_limit()`) doesn't do anything except change an integer limit, while incurring the cost of a string comparison and allocation.

In other words, trying to manage the sizes of the individual pfifos is not needed because they are always exactly as big as is necessary. We can remove all of the code that tries to change their size dynamically (including all of the "borrowing" code), as well as the code that tries to treat priorities 0 and 1 differently.

This should simplify the code quite a bit -- please include it in your code ASAP so we can all review it and you can still have time to test your code and evaluate it before the approaching the deadline.

cjdoucette commented:
By the way, you can set the limit of each pfifo to be the overall maximum limit of the GK priority queue.

ndev2 commented:
Right, I understand. Thanks for the help Cody. I'll update the code I have now, removing the 'set_limits', and rebase it here.

ndev2 commented:
Hi Cody, I've pushed the updated GK module with changes from the code review. Hope the code is good, thanks. 

I've also tested this module and it works fine, passes the stress tests and the iperf evaluations come out with better results than the previous versions.

cjdoucette commented:
Can you squash these commits into one so that we can more easily review the code?

No need to close this pull request and open another. Just do the squash, and once the code is in one commit just do:

$ git push origin Gatekeeper_2 -f

ndev2 commented:
Done, you should be able to see it as one commit now

cjdoucette commented:
Great, we'll take a look. While we do that, try to add the suggestions to the iproute2 repo.

AltraMayor commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,316 @@
+		q->prio_array[prio] = pfifo_q;
+	}
+
+	q->highest_prio = 0; // Set default highest priority to 0.
```
Style: The C99-style "// ..." comments are not allowed in the kernel (see https://github.com/AltraMayor/XIA-for-Linux/blob/5f722839c166daafdf3f4c3c6acd7ae0c8401925/Documentation/CodingStyle#L464 ). Please use `/* ... */`.

AltraMayor commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,316 @@
+		if ((q->prio_array[hp])->q.qlen > 0)
+			return skb;
+
+		q->highest_prio = calc_new_high_prio(q); // update highest_prio
```
Style: use `/* ... */` instead of `//`.

AltraMayor replied:
Style: Invert the condition at the `if` statement above to move this line under the `if` statement, and thus being able to drop the current `return skb;` under that `if` statement.

AltraMayor commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,316 @@
+
+/* Priority markers */
+#define MAX_PRIORITY 64
+#define HALF_MAX_PRIORITY 32
```
`HALF_MAX_PRIORITY` is no longer used, drop it.

AltraMayor commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,316 @@
+
+struct gkprioq_sched_data {
+/* Parameters */
+	u32 max_limit;
```
Indent the comment to align with field `max_limit`.

AltraMayor commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,316 @@
+
+/* Variables */
+	struct Qdisc *prio_array[MAX_PRIORITY];
+	u16 highest_prio;
```
1. The word "Variables" is very general. The expression "Queue state" is more precise.

2. Indent the comment to align with field `max_limit`.

AltraMayor commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,316 @@
+		if (ret != NET_XMIT_SUCCESS) {
+			if (net_xmit_drop_count(ret))
+				qdisc_qstats_drop(sch);
+		return ret;
```
Style: correct indentation.

AltraMayor commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,316 @@
+
+	else {
+		prio = 0;			
+		while (prio < your_prio) { 
```
You can avoid this loop adding field `lowest_prio` in `struct gkprioq_sched_data` and maintaining it like you already do with field `highest_prio`.

During normal operations, Gatekeeper packets will have priorities above a given constant, for example, 30. So this loop will be expensive.

ndev2 replied:
Ok, understood.

AltraMayor commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,316 @@
+	 * and then enqueue the ingressed packet.
+	 */
+
+	else if (your_queue->q.qlen == your_queue->limit || your_prio == 0) {
```
1. If variable `your_prio` is not zero, this condition is possibly dropping a packet whose priority is not the lowest. This is a bug. The solution is to drop this whole `else if`.

2. Style: Given that the `if` above terminates with a `return` and this whole `else if` is going to be dropped, you don't need the `else` below, so remove the indentation of the body of the `else`.

3. Notice that the code in the `else` won't have trouble with condition `your_queue->q.qlen == your_queue->limit` because it will drop a packet before it adds the new packet.

ndev2 replied:
Hi Michel,  yes I was not confident with `else if` because I didn't know if I should drop the ingress packet or a packet already enqueued, before enqueueing another packet. Will make the change.

AltraMayor commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,316 @@
+	u32 max_limit;
+
+/* Variables */
+	struct Qdisc *prio_array[MAX_PRIORITY];
```
While we can accept using `Qdisc` for GSoC, we won't endorse this version of the code if you intend to merge it upstream. Using `struct sk_buff_head` here is a better choice.

ndev2 replied:
Right, understood. Will begin to change the code after GSoC.

AltraMayor reviewed (changes requested):
**Removed 'borrowing' and 'fifo_set_limit's, updated GK module after co…**

AltraMayor commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,316 @@
+				}
+
+				qdisc_qstats_backlog_inc(sch, skb);
+				sch->q.qlen++;
```
Given that a packet was dropped before adding the new packet, do we need to increment `sch->q.qlen`?

ndev2 replied:
We shouldn't, `__qdisc_drop_head` doesn't update the stats on its own, so yes, I've removed the `sch->q.qlen++`  now, thanks.

AltraMayor reviewed (changes requested):
**Removed 'borrowing' and 'fifo_set_limit's, updated GK module after co…**

ndev2 commented:
Hi guys, I've rebased the code after Michel's review, hope the code is good, thanks.

sachinparyani commented:
Nishanth, can you update the module with the updated code and check if it
works. Also, please share the results of the new stress tests and iperf
evaluations.

On Tue, Aug 22, 2017 at 1:38 PM, ndev2 <notifications@github.com> wrote:

> Hi guys, I've rebased the code after Michel's review, hope the code is
> good, thanks.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ndev2/net-next/pull/2#issuecomment-324114966>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/APm1whGqWDjMd20dwIbdG2-3tJ3S-ZFVks5sayAPgaJpZM4O72Ty>
> .
>

ndev2 commented:
Hi Sachin, I just rebased the code after Michel's comments today, and I did do the basic hasty testing and it seems to work fine. I'll share the results of the tests done and recorded properly, in a few hours, thanks.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+	struct gkprioq_sched_data *q = qdisc_priv(sch);
+	wlen = skb_network_offset(skb);
+
+	switch(tc_skb_protocol(skb)) {
```
Add a space between `switch` and the parenthesis.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+		if (!pskb_may_pull(skb, wlen))
+			goto drop;
+
+		your_prio = (ipv4_get_dsfield(ip_hdr(skb))) >> 2;
```
Remove parentheses around `ipv4_get_dsfield(ip_hdr(skb))`, since they're unnecessary.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+		if (!pskb_may_pull(skb, wlen))
+			goto drop;
+
+		your_prio = (ipv6_get_dsfield(ipv6_hdr(skb))) >> 2;
```
Remove parentheses around `ipv6_get_dsfield(ipv6_hdr(skb))`, since they're unnecessary.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+				      &(q->prio_array[lp])->q, to_free);
+		qdisc_qstats_drop(sch);
+		
+		/* update lowest priority field. */
```
Capitalize "update."

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+	/* Set default highest and lowest priorities to 0. */
+
+	q->highest_prio = 0;
+	q->lowest_prio = 0;
```
I think this should be initialized to `MAX_PRIORITY - 1`, since otherwise it will never be changed (we can never have a priority less than 0).

ndev2 replied:
Yes, will change it

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+
+	/* Initialise 64 pfifo queues, one for each possible priority. */
+
+	for(prio = 0; prio < MAX_PRIORITY; prio++) {
```
Add a space between `for` and the parenthesis.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+	else {
+		struct tc_gkprioq_qopt *ctl = nla_data(opt);
+
+		if(nla_len(opt) < sizeof(*ctl))
```
Add a space between `if` and the parenthesis.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+
+	if (opt == NULL) 
+		 q->max_limit = qdisc_dev(sch)->tx_queue_len; 
+	else {
```
Style: add curly braces around the `if` block as well since the `else` block requires them.

ndev2 replied:
Shall I just remove the `else` and its curly braces? That should work

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+
+	sch->qstats.backlog = 0;
+	q->highest_prio = 0;
+	q->lowest_prio = 0;
```
Initialize to `MAX_PRIORITY - 1`.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+	u16 hp = q->highest_prio;
+
+	skb = qdisc_dequeue_head(q->prio_array[hp]);
+	if (likely(skb != NULL)) { 
```
You can simplify this to `if (likely(skb)) {`.

cjdoucette replied:
Let's actually flip the logic of this code so that we can just do:

    if (unlikely(!skb))
            return NULL;

And then let the rest of the function be the case where we've found the `skb` and return it in the end.

ndev2 replied:
Ok, will do.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+	
+		if ((q->prio_array[hp])->q.qlen == 0) {
+			q->highest_prio = calc_new_high_prio(q);
+			return skb;
```
I think this `return skb;` should be outside of this inner `if` statement, since if we were able to dequeue an `skb` then we should return it.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+ /*
```
Remove the space at the beginning of this line.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+#include <net/dsfield.h>
+#include <net/inet_ecn.h>
+#include <asm/byteorder.h>
+#include <linux/bitops.h>
```
Make sure all of these headers are really needed. Try commenting out each statement one-by-one and compile the module each time to see if it's really nodded.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+#include <asm/byteorder.h>
+#include <linux/bitops.h>
+
+/* Priority markers */
```
I think you can improve this comment by making it:

`/* Packets are assigned priorities [0, 63] due to the limits of the IP DSCP field. */`

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+
+
+struct gkprioq_sched_data {
+	    /* Parameters */
```
The indentation here is wrong -- it should be just a single tab from the left. And the comment should end with a period.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+	    /* Parameters */
+	u32 max_limit;
+
+            /* Queue state */
```
Same issue here -- the indentation should be just a single tab, and the comment should end with a period.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+	}
+
+	return 0; /* GK queue is empty, return 0 
+		   * (default highest priority setting). */
```
Can you put this comment on one line above the `return 0;`?

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+	}
+	
+	return 0; /* GK queue is empty, return 0 
+		   * (default lowest priority setting). */
```
Can you put this comment on one line above the `return 0;`?

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+			return prio;
+	}
+	
+	return 0; /* GK queue is empty, return 0 
```
Shouldn't the default lowest priority be `MAX_PRIORITY - 1`? So that when we enqueue a packet we update the lowest priority.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+		if (ret != NET_XMIT_SUCCESS) {
+			if (net_xmit_drop_count(ret))
+				qdisc_qstats_drop(sch);
+				return ret;
```
This statement's indentation is wrong -- should be pulled back to the left by one tab.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+ *
+ *
+ *   The Gatekeper standalone codebase is found here: 
+ *	          
```
This line has a bunch of extra whitespace on it -- remove it.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+ *
+ *
+ *
+ *   The Gatekeper standalone codebase is found here: 
```
Almost all of the lines in the above four paragraphs have extra spaces at the end of each line -- remove them.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+		if ((q->prio_array[prio])->q.qlen > 0)
+			return prio;
+	}
+	
```
This line has extra whitespace -- remove it.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+{
+	struct Qdisc *your_queue;
+	int ret, wlen, gk_len;
+	u16 your_prio, lp; 
```
This line has extra whitespace at the end.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+		break;
+
+	case htons(ETH_P_IPV6):
+	 	wlen += sizeof(struct ipv6hdr);
```
This line has a space between the two tabs -- remove it.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+			goto drop;
+
+		your_prio = (ipv6_get_dsfield(ipv6_hdr(skb))) >> 2;
+	      	break;
```
This line has a bunch of spaces in it, but it should just be two tabs.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+		your_prio = (ipv6_get_dsfield(ipv6_hdr(skb))) >> 2;
+	      	break;
+
+	default: 
```
This line has a space at the end of it -- remove it.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+	if (likely(skb != NULL)) { 
+
+		qdisc_qstats_backlog_dec(sch, skb);
+		sch->q.qlen--;
```
Can you put this line to update the queue length above the two lines that update stats? Just to keep the statistics lines together.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+	if (lp != your_prio) { 		
+		
+		__qdisc_queue_drop_head(q->prio_array[lp], 
+				      &(q->prio_array[lp])->q, to_free);
```
Shouldn't we be dropping the last packet in the lowest priority FIFO since that was the most recent packet that was added? In other words, we should be dropping from the tail, not the head.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,314 @@
+	
+drop:
+	qdisc_drop(skb, sch, to_free);
+	return NET_XMIT_SUCCESS | __NET_XMIT_BYPASS;
```
What does `__NET_XMIT_BYPASS` mean, and is this combination of flags appropriate for the case when we drop the packet that was just requested to be enqueued?

ndev2 replied:
`__NET_XMIT_BYPASS` is saying that the packet should try to call this function again because it could result in a different result at that time. Yes, it should work.

cjdoucette reviewed (changes requested):
This review has a lot of small, style-related comments that you should fix. There are a lot of lines that have extra whitespace in them -- I started to leave comments where it's happening, but there's too many spots for me to explain them all. Instead, when you are done making changes and have all changes added, do this:

`$ git diff HEAD^ HEAD`

And scroll through the patch. If you see any lines that are highlighted in red, that means there is extra whitespace there that should be removed.

There are still two significant logical bugs that are making the module incorrect. One is that when we need to drop the lowest priority packet from the priority queue, you are dropping the packet at the head of the lowest priority queue but should be dropping the packet at the tail.

The other logical error is in the dequeue function. When a packet is correctly dequeued but the highest priority doesn't need to be updated, `NULL` is being returned. See the suggestions I gave around that part of the code for how to change that function.

ndev2 commented:
Ok Cody, will do, I'll make the changes and rebase here, thanks.

ndev2 commented:
Hi guys, I've updated the module, removed white spaces and some other functionality changes in enqueue that Cody mentioned, hope the code is good, thanks.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,327 @@
+	int prio;
+
+	if (q->lowest_prio == 0)
+		return 0;
```
I don't think this `if` statement makes sense for this function. If the old lowest priority was 0, then it could be anything above 0.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,327 @@
+	}
+
+	/* GK queue is empty, return MAX_PRIORITY - 1
+	   (default lowest priority setting). */
```
Style: add a `*` before the second line, and move the `*/` onto its own line.

cjdoucette replied:
Or just make it a one line comment.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,327 @@
+		q->lowest_prio = your_prio;
+
+	your_queue = q->prio_array[your_prio];
+	lp = q->lowest_prio;
```
The variable `lp` is not needed until later in the function -- move this line down to line ~168.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,327 @@
+	lp = q->lowest_prio;
+
+	gk_len = sch->q.qlen;
+	if (gk_len < q->max_limit) {
```
You can entirely get rid of the variable `gk_len` and just use `sch->q.qlen` in this if statement, since `gk_len` is only used once.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,327 @@
+	 * accomodated. Otherwise no choice but to drop ingress packet.
+	 */
+
+	lpq = &(q->prio_array[lp])->q;
```
Since `lpq` is only used inside this next if statement, move it into there. You can move its declaration (`struct qdisc_skb_head *lpq;`) from the top of this function into the if statement as well, so that the first line of the if statement is:

    struct qdisc_skb_head *lpq = &(q->prio_array[lp])->q;

cjdoucette commented:
After these changes are made, I think the module is in acceptable condition to test. We might have to work on it later to put it into condition to merge it with the kernel, but I think it is at least correct and efficient now.

ndev2 commented:
Hi guys, made the style changes as Cody suggested, hope the code is good, thanks.

AltraMayor commented:
Thank you for such a careful review, @cjdoucette.

To conclude this review of this part of the project for GSoC, there's one more item to do: addressing the feedback obtained with `checkpatch.pl` (it is available in the kernel source). You should run `checkpatch.pl` on all files you modified. For example, on sch_gkprioq.c I got the following feedback:
```
$ scripts/checkpatch.pl -f net/sched/sch_gkprioq.c
WARNING: Missing a blank line after declarations
#111: FILE: sch_gkprioq.c:111:
+	struct gkprioq_sched_data *q = qdisc_priv(sch);
+	wlen = skb_network_offset(skb);

WARNING: please, no space before tabs
#307: FILE: sch_gkprioq.c:307:
+^I.destroy ^I=^Igkprioq_destroy,$

total: 0 errors, 2 warnings, 324 lines checked

sch_gkprioq.c has style problems, please review.

NOTE: If any of the errors are false positives, please report
      them to the maintainer, see CHECKPATCH in MAINTAINERS.
```

@ndev2, it should be easy to address all feedback `checkpatch.pl` generates. Once you fix all feedback, run `checkpatch.pl` again to make sure that no errors or warnings were introduced with the last changes. 

Congratulations for getting until here. After GSoC, we will start the next level of review, in which we'll make the code ready for an eventual merge upstream.

Don't forget to work on the tc pull request!

ndev2 commented:
@AltraMayor  Thanks Michel, I've updated the module, it now clears `checkpatch.pl`. Thanks mentors @cjdoucette  @AltraMayor  @sachinparyani  for the code review, pleasure working with you all. I'll get started on the `iproute2` repo  right away.

cjdoucette commented on include/uapi/linux/pkt_sched.h:
```diff
@@ -122,6 +122,12 @@ struct tc_fifo_qopt {
 	__u32	limit;	/* Queue length: bytes for bfifo, packets for pfifo */
 };
 
+/* GKPRIOQ section */
```
Let's make the abbreviation for this `GKPRIO` instead of `GKPRIOQ` to more closely match the simple `PRIO` queue.

cjdoucette commented on net/sched/Kconfig:
```diff
@@ -229,6 +229,17 @@ config NET_SCH_MQPRIO
 
 	  If unsure, say N.
 
+config NET_SCH_GKPRIOQ
```
Change to `NET_SCH_GKPRIO`.

cjdoucette commented on net/sched/Kconfig:
```diff
@@ -229,6 +229,17 @@ config NET_SCH_MQPRIO
 	  If unsure, say N.
 
+config NET_SCH_GKPRIOQ
+	tristate "Gatekeeper DoS Defense priority queue scheduler (GKPRIOQ)"
```
Let's just leave this as `"Gatekeeper priority queue scheduler"`.

cjdoucette commented on net/sched/Kconfig:
```diff
@@ -229,6 +229,17 @@ config NET_SCH_MQPRIO
+config NET_SCH_GKPRIOQ
+	tristate "Gatekeeper DoS Defense priority queue scheduler (GKPRIOQ)"
+	help
+	  Say Y here if you want to use the Gatekeeper Priority Queue scheduler.
```
Make "Priority" and "Queue" lowercase here.

cjdoucette commented on net/sched/Kconfig:
```diff
@@ -229,6 +229,17 @@ config NET_SCH_MQPRIO
+	tristate "Gatekeeper DoS Defense priority queue scheduler (GKPRIOQ)"
+	help
+	  Say Y here if you want to use the Gatekeeper Priority Queue scheduler.
+	  This scheduler is useful for DoS mitigation.
```
Let's make the description:

"This schedules packets according to priorities based on the DSCP (IPv4) and DS (IPv6) fields, which is useful for request packets in DoS mitigation systems such as Gatekeeper."

cjdoucette commented on net/sched/Kconfig:
```diff
@@ -229,6 +229,17 @@ config NET_SCH_MQPRIO
+	  This scheduler is useful for DoS mitigation.
+
+	  To compile this driver as a module, choose M here: the module will
+	  be called sch_gkprioq.
```
Change to `sch_gkprio`.

cjdoucette commented on include/uapi/linux/pkt_sched.h:
```diff
@@ -122,6 +122,12 @@ struct tc_fifo_qopt {
 
+/* GKPRIOQ section */
+
+struct tc_gkprioq_qopt {
```
Change to `struct tc_gkprio_qopt`.

cjdoucette commented on net/sched/Makefile:
```diff
@@ -45,6 +45,7 @@ obj-$(CONFIG_NET_SCH_NETEM)	+= sch_netem.o
 obj-$(CONFIG_NET_SCH_DRR)	+= sch_drr.o
 obj-$(CONFIG_NET_SCH_PLUG)	+= sch_plug.o
 obj-$(CONFIG_NET_SCH_MQPRIO)	+= sch_mqprio.o
+obj-$(CONFIG_NET_SCH_GKPRIOQ)	+= sch_gkprioq.o
```
Change abbreviations on both sides.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+/*
+ * net/sched/sch_gkprioq.c  Gatekeeper Priority Queue.
```
Change filename abbreviation.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+ *
+ * Authors:	Nishanth Devarajan, <ndev_2021@gmail.com>
+ *	        original idea by Michel Machado, Cody Doucette and Qiaobin Fu
+ *
```
Remove this blank line, and add a comma after "Doucette" above.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+/* Packets are assigned priorities [0, 63] due to the IP DSCP field limits. */
+#define MAX_PRIORITY 64
+
+
```
Remove this blank line.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+/*	  Gatekeeper Priority Queue
+ *	=================================
+ *
+ *
```
Remove this blank line.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+ *
+ *
+ * Gatekeeper is an open source DoS defense software which sends traffic based
+ * on an assigned priority for the client's requeust. Assigning priorities to
```
Typo on "request."

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+ *
+ *          https://github.com/AltraMayor/gatekeeper
+ *
+ *
```
Remove these two blank lines. And on the lines above, leave just a single blank line between paragraphs.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+ *
+ */
+
+
```
Remove this blank line.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+ */
+
+
+struct gkprioq_sched_data {
```
Change to `struct gkprio_sched_data`.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+	/* GK queue is empty, return MAX_PRIORITY - 1
+	 * (default lowest priority setting).
+	 */
+
```
Remove this blank line.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+	your_queue = q->prio_array[your_prio];
+
+	if (sch->q.qlen < q->max_limit) {
+
```
Remove this blank line.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+
+		sch->q.qlen++;
+		return __qdisc_enqueue_tail(skb, sch, your_queue);
+
```
Remove this blank line.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+
+	lp = q->lowest_prio;
+	if (lp != your_prio) {
+
```
Remove this blank line.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+
+		pkt = lpq->tail;
+
+		if (likely(pkt != NULL)) {
```
Make this `if (likely(pkt)) {`.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+		if (likely(pkt != NULL)) {
+			lpq->tail = pkt->prev;
+			lpq->qlen--;
+			if (lpq->tail == NULL)
```
Make this `if (!lpq->tail)`.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+			q->lowest_prio = calc_new_low_prio(q);
+
+		return __qdisc_enqueue_tail(skb, sch, your_queue);
+
```
Remove this blank line.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+		qdisc_qstats_drop(sch);
+
+		/* Update lowest priority field. */
+
```
Remove this blank line.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+		/* Drop the packet at the tail of lowest_priority queue. */
+
+		pkt = lpq->tail;
+
```
Remove this blank line.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+		struct qdisc_skb_head *lpq = q->prio_array[lp];
+
+		/* Drop the packet at the tail of lowest_priority queue. */
+
```
Remove this blank line.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+	}
+
+	return qdisc_drop(skb, sch, to_free);
+
```
Remove this blank line.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+	qdisc_bstats_update(sch, skb);
+
+	/* Update highest priority field. */
+
```
Remove this blank line.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+	int prio;
+	struct gkprioq_sched_data *q = qdisc_priv(sch);
+
+	if (opt == NULL) {
```
Make this `if (!opt) {`.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+
+	if (opt == NULL) {
+		q->max_limit = qdisc_dev(sch)->tx_queue_len;
+
```
Remove this blank line.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+	}
+
+	/* Initialise 64 queues, one for each possible priority. */
+
```
Remove this blank line.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+	for (prio = 0; prio < MAX_PRIORITY; prio++)
+		qdisc_skb_head_init(q->prio_array[prio]);
+
+
```
Remove this blank line.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+
+
+	/* Set default highest and lowest priorities to 0. */
+
```
Remove this blank line, and remove this comment. It's obvious what the code below is doing.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+	if (nla_put(skb, TCA_OPTIONS, sizeof(opt), &opt))
+		goto nla_put_failure;
+	return skb->len;
+
```
Move this blank line to be before the `return` statement above, not after.

cjdoucette commented on net/sched/sch_gkprioq.c:
```diff
@@ -0,0 +1,308 @@
+
+	for (prio = 0; prio < MAX_PRIORITY; prio++)
+		__qdisc_reset_queue(q->prio_array[prio]);
+
```
Remove this blank line.

cjdoucette reviewed (changes requested):
A lot of small stylistic changes here to put it into shape for the kernel. Be sure to rename the file net/sched/sch_gkprioq.c to net/sched/sch_gkprio.c (withou the "q").

I will continue to do a review of two more things: the comment at the top of the source file to make sure we're describing why this is needed for Gatekeeper in the best way, and the basic algorithm/data structures to make sure everything is correct.

Have you run any tests on the code since changing it? It would be really helpful if you could run the same kind of tests that we did before -- with various priorities, making sure that higher priority packets are being serviced first. And that the lowest priority packets are correctly being dropped when the queue is full.

ndev2 commented:
Hi Cody, I've made the changes as requested, hope its good. I haven't thoroughly tested the new code as was done before, will get that done by tomorrow and send you an email about what I find.

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,288 @@
+#include <net/inet_ecn.h>
+
+/* Packets are assigned priorities [0, 63] due to the IP DSCP field limits. */
+#define MAX_PRIORITY 64
```
Let's make this `GKPRIO_MAX_PRIORITY`.

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,288 @@
+	const struct qdisc_skb_head *temp_queue;
+
+	for (prio = q->highest_prio - 1; prio >= q->lowest_prio; prio--) {
+		temp_queue = &q->prio_array[prio];
```
Since `temp_queue` is only used inside the `for` loop, let's remove the declaration of it above and instead put it here (dropping the `const` too): `struct qdisc_skb_head *temp_queue = &q->prio_array[prio];`.

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,288 @@
+	u32 max_limit;
+
+	/* Queue state. */
+	struct qdisc_skb_head prio_array[MAX_PRIORITY];
```
Let's rename this array `qdiscs`: `struct qdisc_skb_head qdiscs[GKPRIO_MAX_PRIORITY];`

This will require changing uses of `prio_array` in the rest of the file to `qdiscs`.

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,288 @@
+	struct gkprio_sched_data *q = qdisc_priv(sch);
+	struct sk_buff *skb;
+	u16 hp = q->highest_prio;
+
```
Delete this blank line.

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,288 @@
+{
+	struct gkprio_sched_data *q = qdisc_priv(sch);
+	struct sk_buff *skb;
+	u16 hp = q->highest_prio;
```
Since we only use `hp` once, we don't need a separate variable for it. Just use the expression `q->highest_prio` directly in the line below.

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,288 @@
+static inline u16 calc_new_low_prio(const struct gkprio_sched_data *q)
+{
+	int prio;
+	const struct qdisc_skb_head *temp_queue;
```
Same comment as above -- we can move the declaration of this to inside the `for` loop, and drop the `const`.

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,288 @@
+static int gkprio_enqueue(struct sk_buff *skb, struct Qdisc *sch,
+			   struct sk_buff **to_free)
+{
+	struct qdisc_skb_head *your_queue;
```
Let's rename this `qdisc`: `struct qdisc_skb_head *qdisc;`

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,288 @@
+			   struct sk_buff **to_free)
+{
+	struct qdisc_skb_head *your_queue;
+	struct sk_buff *pkt;
```
Let's rename this `to_drop`: `struct sk_buff *to_drop;`

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,288 @@
+	struct qdisc_skb_head *your_queue;
+	struct sk_buff *pkt;
+	int wlen;
+	u16 your_prio, lp;
```
Let's rename `your_prio` to just `prio`.

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,288 @@
+
+		your_prio = ipv4_get_dsfield(ip_hdr(skb)) >> 2;
+		break;
+
```
Delete this blank line.

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,288 @@
+
+		your_prio = ipv6_get_dsfield(ipv6_hdr(skb)) >> 2;
+		break;
+
```
Delete this blank line.

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,288 @@
+		q->highest_prio = your_prio;
+
+	if (your_prio < q->lowest_prio)
+		q->lowest_prio = your_prio;
```
If we end up dropping this packet because the queue is already full, then this update is wrong. We should only update the highest priority and lowest priority if this packet actually gets added to the queue and is the highest or lowest. Therefore, let's move this code that updates these values into the `if` statement below, just before incrementing `qlen` and then enqueuing the packet.

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,288 @@
+		return __qdisc_enqueue_tail(skb, sch, your_queue);
+	}
+
+	/* Look to lowest priority queue to check if ingress packet can be
```
The changes to the rest of the function are hard to describe, as I ended up rewriting it a little bit. So here's how the rest of this function should go:

    	/* If this packet is lowest priority, drop it. */
    	lp = q->lowest_prio;
    	if (prio <= lp)
    		return qdisc_drop(skb, sch, to_free);

    	/* Drop the packet at the tail of lowest priority qdisc. */
    	lp_qdisc = &q->qdiscs[lp];
    	to_drop = lp_qdisc->tail;
    	if (likely(to_drop)) {
    		lp_qdisc->tail = to_drop->prev;
    		lp_qdisc->qlen--;
    		if (!lp_qdisc->tail)
    			lp_qdisc->head = NULL;
    		to_drop->prev = NULL;
    	}
    	__qdisc_drop(to_drop, to_free);
    	qdisc_qstats_drop(sch);

    	ret = __qdisc_enqueue_tail(skb, sch, qdisc);

    	/* Check to update highest and lowest priorities. */
    	if (lpq->qlen == 0)
    		q->lowest_prio = calc_new_low_prio(q);

    	if (prio > q->highest_prio)
    		q->highest_prio = prio;

    	return ret;
    drop:
    	qdisc_drop(skb, sch, to_free);
    	return NET_XMIT_SUCCESS | __NET_XMIT_BYPASS;

To do this, you'll need to add these two declarations to the beginning of the function:

    int ret;
    struct qdisc_skb_head *lp_qdisc;

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,288 @@
+	return 0;
+}
+
+static int gkprio_dump(struct Qdisc *sch, struct sk_buff *skb)
```
Shouldn't we be providing more information about the state of the GK priority queue in this function, not just its size limit?

ndev2 commented:
HI Cody, I've made the requested changes, hope they're good. In regards to the 'dump' function, currently I've changed it to show highest and lowest prios in GK, let me know if we want to show more stats (I was thinking no. of non empty queues or the number of pkts queued in each queue..  but  let me know, just in case) I'd have to update tc code to reflect these dump stats, but was planning to do that after we've finalised the stats we want to show... it shouldn't be huge trouble to update the tc space in any case. Thanks.

cjdoucette commented on include/uapi/linux/pkt_sched.h:
```diff
@@ -122,6 +122,16 @@ struct tc_fifo_qopt {
+
+	/* Stats */
+	__u16 highest_prio; /* Highest priority currently in queue  */
+	__u16 lowest_prio;  /* Lowest priority currently in queue */
```
Add periods to the end of these three comments.

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,286 @@
+	
+	for (prio = q->highest_prio - 1; prio >= q->lowest_prio; prio--) {
+		struct qdisc_skb_head *temp_queue = &q->qdiscs[prio];
+		if (temp_queue->qlen > 0)
```
Since we only use `temp_queue` once, I think we can get rid of the variable altogether. Just use the expression `q->qdiscs[prio].qlen` in the `if` statement.

Same thing applies for the function `calc_new_low_prio()` below.

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,286 @@
+
+	if (sch->q.qlen < q->max_limit) {
+
+	/* Check to update highest and lowest priorities. */
```
Push this comment in by one indent.

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,286 @@
+			q->lowest_prio = prio;
+
+		sch->q.qlen++;
+		return __qdisc_enqueue_tail(skb, sch, qdisc);
```
`__qdisc_enqueue_tail()` always succeeds, but it has a return value, so error conditions could be added in the future. If in the future it can fail, then we need to make sure it succeeded first before changing the state of our queue.

This `if` statement should therefore be:

    if (sch->q.qlen < q->max_limit) {
    	ret = __qdisc_enqueue_tail(skb, sch, qdisc);
    	if (ret < 0)
    		return ret;

    	/* Check to update highest and lowest priorities. */
    	if (prio > q->highest_prio)
    		q->highest_prio = prio;
                
    	if (prio < q->lowest_prio)
    		q->lowest_prio = prio;
                        
     	sch->q.qlen++;
    	return ret;
    }

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,286 @@
+	struct sk_buff *to_drop;
+	int wlen, ret;
+	u16 prio, lp;
+	struct gkprio_sched_data *q = qdisc_priv(sch);
```
Since we don't actually use `q` until after the `switch` statement, let's only declare `q` here but initialize it just before initializing `qdisc` after the `switch`.

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,286 @@
+		to_drop->prev = NULL;
+	}
+	__qdisc_drop(to_drop, to_free);
+	qdisc_qstats_drop(sch);
```
This functionality is essentially what a function called `qdisc_queue_drop_tail()` would do if it existed, but the qdisc library only has a function `qdisc_queue_drop_head()`. Let's add `qdisc_queue_drop_tail()` to `net/sched/sch_generic.h` using `qdisc_queue_drop_head()` as an example.

Then, replace the code in this part of the function with a call to `qdisc_queue_drop_tail()`.

Doing this will greatly simplify the code here in our enqueue function, and will contribute another function to the kernel that others may be able to use.

cjdoucette commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,286 @@
+ 
+	opt.limit = q->max_limit;
+	opt.highest_prio = q->highest_prio;
+	opt.lowest_prio = q->lowest_prio;
```
Do other schedulers provide more information, such as information about how many packets of each priority there are? For example, what does the prio scheduler do?

We should essentially be looking for what other schedulers do and try to do something similar.

ndev2 replied:
Ok, after looking around, I find that all the schedulers do not provide too much information of their state, and none of them talk in too much  depth of the packets they're handling (prio shows shows the number of bands its tracking and a map from the possible logical priorities to their corresponding band, this is a lot less than what they can show).. so this should be good (I'm more than 90% sure)

cjdoucette commented on include/net/sch_generic.h:
```diff
@@ -742,6 +742,57 @@ static inline struct sk_buff *qdisc_peek_head(struct Qdisc *sch)
+	return skb;
+}
+
+static inline struct sk_buff *qdisc_dequeue_tail(struct Qdisc *sch)
```
Since we are not using `qdisc_dequeue_tail()`, let's not include it in this patch. We don't want to add code to the kernel that doesn't have any callers. If someone needs to use this version of the function (and not `__qdisc_dequeue_tail()`), they can add it in the future.

There is a `qdisc_dequeue_head()` because there is code elsewhere in the kernel that calls it.

cjdoucette commented on include/uapi/linux/pkt_sched.h:
```diff
@@ -122,6 +122,16 @@ struct tc_fifo_qopt {
+struct tc_gkprio_qopt {
+	__u32	limit;	/* Queue length in packets. */
+
+	/* Stats */
```
Add a period to this comment.

cjdoucette commented on include/net/sch_generic.h:
```diff
@@ -742,6 +742,57 @@ static inline struct sk_buff *qdisc_peek_head(struct Qdisc *sch)
+	return __qdisc_queue_drop_tail(sch, &sch->q, to_free);
+}
+
+
```
Remove this blank line.

cjdoucette reviewed (commented):
Once you do these simple fixes, my review of the code is complete!

Please also make the commit subject `net/sched: add gkprio scheduler`.

Let me know once that's done, and I'll tell Michel that he can review the code.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,283 @@
+ * (IPv4) or DS (IPv6) field, where a higher value places the packet closer
+ * to the exit of the queue. Non-IP packets are assigned priority 0. When the
+ * queue is full, the lowest priority packet in the queue is dropped to make
+ * room for the next packet to be added.
```
Let's be precise here (updating the last line of the comment and adding two more lines):
```
 * room for the packet to be added if it has higher priority.
 * If the packet to be added has lower priority than all packets in the queue,
 * it is dropped.
```

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,283 @@
+ *
+ * This qdisc schedules a packet according to the value (0-63) of its DSCP
+ * (IPv4) or DS (IPv6) field, where a higher value places the packet closer
+ * to the exit of the queue. Non-IP packets are assigned priority 0. When the
```
Since we are proposing a queue discipline that hopefully will be useful elsewhere, not only for Gatekeeper, we should parametrize the priority that non-IP packets receive in the same way we get the maximum size of the queue.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,283 @@
+	u32 max_limit;
+
+	/* Queue state. */
+	struct qdisc_skb_head qdiscs[GKPRIO_MAX_PRIORITY];
```
Why did we choose `struct qdisc_skb_head` instead of `struct sk_buff_head`? These structs are basically identical, but the following comment, which is found just above the definition of `struct qdisc_skb_head` in file `include/net/sch_generic.h`, points out that pointer `skb->prev` is not expected to be defined for `struct qdisc_skb_head`:
`/* similar to sk_buff_head, but skb->prev pointer is undefined. */`

This is important because `__qdisc_dequeue_tail()`, for example, requires pointer `skb->prev` to work.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,283 @@
+	u16 lowest_prio;
+};
+
+static inline u16 calc_new_high_prio(const struct gkprio_sched_data *q)
```
This function is "too long" for the tab `inline`. Please drop the tag `inline`.

By the way, since this function is `static` and used only once, the compiler is going to make it `inline` implicitly. But dropping the tag `inline` here for a function that is "too long" should reduce comments during the mainline review.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,283 @@
+	return 0;
+}
+
+static inline u16 calc_new_low_prio(const struct gkprio_sched_data *q)
```
This function is "too long" for the tab `inline`. Please drop the tag `inline`.

By the way, since this function is `static` and used only once, the compiler is going to make it `inline` implicitly. But dropping the tag `inline` here for a function that is "too long" should reduce comments during the mainline review.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,283 @@
+
+	wlen = skb_network_offset(skb);
+
+	switch (tc_skb_protocol(skb)) {
```
Style: drop the bank line between the initialization of `wlen` and the `switch` statement above, and add the following comment before the initialization of `wlen`:
`/* Obtain the priority of @skb. */

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,283 @@
+	opt.lowest_prio = q->lowest_prio;
+
+	if (nla_put(skb, TCA_OPTIONS, sizeof(opt), &opt))
+		goto nla_put_failure;
```
Style: there is only one label (i.e. `nla_put_failure`) and it's only used here, so just `return -1;` here and drop the `goto` and label.

AltraMayor commented on include/net/sch_generic.h:
```diff
@@ -742,6 +742,44 @@ static inline struct sk_buff *qdisc_peek_head(struct Qdisc *sch)
+}
+
+static inline unsigned int qdisc_queue_drop_tail(struct Qdisc *sch,
+						 struct sk_buff **to_free)
```
Given that we are introducing a function analog to `qdisc_queue_drop()`, we should keep the parallels between the functions as true as possible. For example, `qdisc_queue_drop()` does not require parameter `struct sk_buff **to_free`. Why do we need to do so here?

ndev2 replied:
I was making it parallel to `qdisc_queue_drop_head()` which uses `to_free` as a parameter. I just thought this function would be more aligned to the function we're writing here, hope that's ok.

AltraMayor replied:
We should drop `qdisc_queue_drop_tail()` all together because our code is not using it. Having a new function that is not called is a sure thing to receive comments during the review process.

AltraMayor commented on include/net/sch_generic.h:
```diff
@@ -742,6 +742,44 @@ static inline struct sk_buff *qdisc_peek_head(struct Qdisc *sch)
 	return qh->head;
 }
 
+static inline struct sk_buff *__qdisc_dequeue_tail(struct qdisc_skb_head *qh)
```
The function `__qdisc_dequeue_tail()` already exists in the kernel: https://elixir.bootlin.com/linux/v4.5/source/include/net/sch_generic.h#L644

Moreover, the kernel version of `__qdisc_dequeue_tail()` works on `struct sk_buff_head` instead of on `struct qdisc_skb_head`, what constitutes further support to replace our use of `struct qdisc_skb_head` in favor of `struct sk_buff_head` in `struct gkprio_sched_data`. See my comment at `struct gkprio_sched_data` for more information.

ndev2 replied:
Hi Michel, Im working on the changes, and one thing I came across: the current linux and net-next repo: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/net/sch_generic.h
do not have the`__qdisc_dequeue_tail()` function... they removed it in one of the v4.8 rcs.  Being the case, I'm not completely sure if we should incorporate the function in the patch (after it being removed once already), any thoughts on this?

AltraMayor replied:
Hi @ndev2,

I've found the exact commit that drops `__qdisc_dequeue_tail()`: `a09ceb0e08140a1eec05b49b4c232d3481339cb0`. You can see what it does with the following command: `git diff -r a09ceb0e08140a1eec05b49b4c232d3481339cb0~ -r a09ceb0e08140a1eec05b49b4c232d3481339cb0`. Essentially, it dopped `__qdisc_dequeue_tail()` because after the changes the commits makes, there were no longer callers to `__qdisc_dequeue_tail()`.

Nevertheless, the function `__skb_dequeue_tail()`, which `__qdisc_dequeue_tail()` has used, is still available in the kernel. Thus, you don't need to reintroduce `__qdisc_dequeue_tail()`, but can take advantage of the previous code of `__qdisc_dequeue_tail()` to implement the code we need.

ndev2 replied:
Right ok. But if we're incorporating this function again, should'nt we try to make it parallel to `__qdisc_dequeue_head()`? and that function uses `qdisc_skb_head` as the parameter. My thinking in using `qdisc_skb_head` instead of `sk_buff_head` is that they've provided API just for our cases like ours, and making use of it would be better practise. Let me know what you think.

AltraMayor replied:
I don't know why they have decided to make `skb->prev` undefined with `struct qdisc_skb_head`, but without `skb->prev`, we can't have a function that drops at the tail. That's why we need to go with `struct sk_buff_head`.

Given that the code of `__qdisc_dequeue_head()` is small and we only need it once in `gkprio_dequeue()`, we can just use the code of `__qdisc_dequeue_head()` directly in `gkprio_dequeue()`. This would make the patch simpler for the kernel reviewers.

ndev2 replied:
Hi Michel, skb->prev is still defined, it is part of the `struct sk_buff` construct and not part of the list construct. I think what they mean is that the reference to the previous `sk_buff` which appeared before the start of the `sk_buff_head` list is no longer maintained (`sk_buff_head` is written with an idea to be able to represent and maintain multiple `sk_buff` lists with operations to support splicing, joining of lists, etc. While `qdisc_skb_head` is written to represent a single queue of `sk_buff`, please take a look at these links to:  https://elixir.bootlin.com/linux/v4.16-rc5/source/include/linux/skbuff.h    https://elixir.bootlin.com/linux/v4.16-rc5/source/include/net/sch_generic.h )  Our code in `gkprio_enqueue()` is currently able to drop packets at the tail of the list.

AltraMayor reviewed (changes requested):
**net/sched: add gkprio scheduler**

Once we are done with my review, you'll be ready to submit a patch to the kernel. Then, you'll want to follow closely the advice in the following page:

Submitting patches: the essential guide to getting your code into the kernel
https://www.kernel.org/doc/html/v4.12/process/submitting-patches.html

Moving toward those advices, add the following lines to the end of the message of your patch:
```
Signed-off-by: Nishanth Devarajan <ndev_2021@gmail.com>
Reviewed-by: Sachin Paryani <sachin.paryani@gmail.com>
Reviewed-by: Cody Doucette <doucette@bu.edu>
Reviewed-by: Michel Machado <michel@digirati.com.br>
```
This will make us comply with items 11, 12, and 13 of the page above as well as show that you're not working alone.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,298 @@
+	u16 noip_dfltp;
+
+	/* Queue state. */
+	struct qdisc_skb_head qdiscs[GKPRIO_MAX_PRIORITY];
```
You haven't replaced `struct qdisc_skb_head` with `struct sk_buff_head` yet.

ndev2 replied:
Hi Michel, I'm really not sure whether we should replace the data structure here. No other net/sched scheduler seems to be using `sk_buff_head` while many schedulers use `qdisc_skb_head` and it even being defined in `sch_generic.h`, I think its strongly suggested that we use `qdisc_skb_head` instead of `sk_buff_head` (please look at the other comment I made before this). Let me know what you think. Thanks.

AltraMayor replied:
Is there a scheduler that uses `struct qdisc_skb_head` **and** adds to the tail of the list? Since that the current kernel doesn't have a function to add to the tail of `struct qdisc_skb_head`, finding that example would clarify what we could do. I don't think there's such example, though. If you are still not convinced of the need to change, you can try to find a scheduler that adds to the tail, and see what `struct` it uses.

ndev2 replied:
I looked through the schedulers and yes, htb uses the `struct qdisc_skb_head`  :  'direct queue' to add to the tail using its static `htb_enqueue_tail` function.

AltraMayor replied:
I made a mistake in my previous comment. What I meant by "and adds to the tail of the list" was really "and drops at the tail of the list" since adding to the tail doesn't require the previous node pointer, but dropping at the tail of a list, as we need in our scheduler, requires. `htb_enqueue_tail()`, which you pointed out, doesn't update the previous pointer for example.

ndev2 replied:
Hi Michel, as I mentioned in another comment the `skb->prev` is still defined, it is inherently a part of the `sk_buff` construct and not the `qdic_skb_head` construct. Our code currently does drop at the tail: loc 151 which is the code to drop packets at the tail of the list in the enqueue function, updating the prev ptr in the process. 

This functionality was  tested in one of the code reviews with Cody, and it did work functionally well (expected good results were obtained in those tests). Let me know what you think. Thanks.

AltraMayor replied:
@ndev2, my goal is to make your code as well aligned with the kernel as possible as to minimize the chance of rejection during the review phase. The more your code reconstructs code present elsewhere or digresses from what is expected from it, the more you're going to hear complains during the review.

On the other hand, I feel I can't make it as clear to you as it is to me. I don't want to force you to do what you don't agree or understand, so I think you should pursue the merge upstream using your understanding how to do so. Good luck.

ndev2 replied:
Ok, I was not clear on what was meant by making `skb->prev`  'undefined'. I apologise for the misunderstanding, and I will make the change to `sk_buff_head` as I understand that you're not allowed to update `skb->prev` in `qdisc_skb_head` code. Thank you for your comments. Hope you can look past my mistake, and we can work together on this.

AltraMayor replied:
@ndev2, I want to work with you, but you must commit yourself to teamwork. If teamwork is not your thing, you may want to give up on coding now while you still have time to find and pursue a solitary activity. Because not being a good team player is going to limit yourself to small pieces of software.

I'm not abruptly saying these things. Not only is this topic quite old -- we have dissussed it back in August 2017 https://github.com/ndev2/net-next/pull/2#discussion_r134526042 -- it is the second time that I point out a problem in your code that must be addressed and, instead of trying to understand my point and move on, you come up with unsound arguments and waste everyone's time. The first time was when I pointed out an invalid memory access in an early version of this scheduler, and you insisted that it was correct justifying that it didn't triggered an error during tests.

I'm not saying this to make you accept whatever I say blindly because this behavor would ultimately doom our project. Nevertheless, if you stick with us, you have to personally grow to make it worth it for everyone, including yourself.

ndev2 replied:
Yes, I understand what you mean, it is a an issue that I've discovered only recently, but it is apparent that its been a problem for a long time now. I'm working on it and trying to be better (outside of coding as well). I'm grateful for your patience, will work on the code to make it good for the kernel. Thanks.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,298 @@
+			return -EINVAL;
+
+		q->max_limit = ctl->limit;
+		q->noip_dfltp = ctl->dfltp;
```
1. The code must check that `ctl->limit` and `ctl->dfltp` have valid numbers.

2. Both parameters must be optional, that is, any of them can be missing. You can use the value with all bits on (i.e. -1) to mark a parameter is missing.

ndev2 replied:
`ctl->limit` checking is automatically done by code outside the queuing discipline, and I've left checking for `ctl->dfltp` to be done in userspace since we know the possible priorities that are allowed,  hope this is ok.

AltraMayor replied:
Kernel code cannot rely on userspace code to sanitize its values. The reason for this rule is that someone can write another userspace application that tries to trick the kernel sending unsanitized values. All kernel code must make sure that whatever it gets from userland is safe before using the information.

This safety verification may lead to duplicate sanitation code: one in the kernel and another in userland applications. But there's a difference here. The code in the kernel is required and essential to keep the safety of the system, whereas the code in the userland application is there mostly to provide easier to understand error messages.

AltraMayor reviewed (changes requested):
**net/sched: add gkprio scheduler**

Please write a final message for this commit according to the link I gave in the previous review, so I can give you feedback on this message as well.

ndev2 commented:
Hi everyone, I was hoping to get some help on the commit message, not really sure what has to go in (in terms of functionality) and what is not necessary, any ideas would be helpful, thanks.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,310 @@
+		if (nla_len(opt) < sizeof(*ctl))
+			return -EINVAL;
+
+		if (ctl->limit == -1) {
```
Replacing `-1` with `(typeof(ctl->limit))-1` avoids compiler warnings like this one: warning: comparison between signed and unsigned integer expressions.

Do the analogous change for the case `ctl->dfltp` below.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,310 @@
+
+		if (ctl->limit == -1) {
+			q->max_limit = qdisc_dev(sch)->tx_queue_len;
+			q->noip_dfltp = ctl->dfltp;
```
We must not copy `q->noip_dfltp` from `ctl->dfltp` here because a malicious application could send `-1` in both parameters, and the code would assign `-1` to `q->noip_dfltp` instead of 0.

Do the analogous change for the case `ctl->dfltp` below.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,310 @@
+		if (ctl->limit == -1) {
+			q->max_limit = qdisc_dev(sch)->tx_queue_len;
+			q->noip_dfltp = ctl->dfltp;
+		}
```
Use an `else if` here to test if the passed parameter is valid.

Do the analogous change for the case `ctl->dfltp` below.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,310 @@
+			q->noip_dfltp = ctl->dfltp;
+		}
+
+		else if (ctl->dfltp == -1) {
```
Make this `if` statement independent of the previous one.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,310 @@
+		else {
+			q->max_limit = ctl->limit;
+			q->noip_dfltp = ctl->dfltp;
+		}
```
Once the changes above are made, this `else` clause won't be needed.

AltraMayor reviewed (changes requested):
**net/sched: add gkprio scheduler//Options checking in init.**

ndev2 commented:
Hi everyone, I've changed the data structure for the qdiscs from `qdisc_skb_head` to `sk_buff_head`, and the relevant changes to the code and the functions called. Locking required for the `sk_buff_head` function is taken care by the `sch_generic.c` code which locks the scheduler when performing qdisc ops.  The options checking in the `init` function now checks for possible erroneous input. Hope the code is good, thanks.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,309 @@
+		if (ctl->limit == (typeof(ctl->limit)) - 1)
+			q->max_limit = qdisc_dev(sch)->tx_queue_len;
+
+		else if (ctl->limit > qdisc_dev(sch)->tx_queue_len) 
```
Style: drop the blank line above the `else if`.

Do analogous change to the similar `if` statement below.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,309 @@
+		else if (ctl->limit > qdisc_dev(sch)->tx_queue_len) 
+			return -EINVAL;
+
+		q->max_limit = ctl->limit;
```
You need to add an `else` on the blank line above, otherwise `q->max_limit` can `-1`.

Do analogous change to the similar `if` statement below.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,309 @@
+		if (ctl->dfltp == (typeof(ctl->dfltp)) - 1) 
+			q->noip_dfltp = 0;
+
+		else if (ctl->dfltp > 63)
```
Replace the test `(ctl->dfltp > 63)` with `(ctl->dfltp >= GKPRIO_MAX_PRIORITY)` to avoid the magic number `63`.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,309 @@
+		break;
+	}
+
+	q = qdisc_priv(sch);
```
Move the assignment of `q` before the `switch` statement; otherwise, there's a bug in `default` case.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,309 @@
+	qdisc = &q->qdiscs[prio];
+
+	if (sch->q.qlen < q->max_limit) {
+		skb_queue_tail(qdisc, skb);
```
I'm under the impression that we can replace `skb_queue_tail()` with `__skb_queue_tail()` here and avoid the lock. Could you verify it? Doesn't the kernel already get a lock for the queue discipline before calling enqueue and dequeue in order to serialize these calls?

ndev2 replied:
Yes it is possible, will make the change... Yes, the kernel holds a lock before performing qdisc ops  in any scheduler inside `sch_generic.c`

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,309 @@
+                lp_qdisc->prev = to_drop->prev;
+                lp_qdisc->qlen--;
+                if (!lp_qdisc->prev)
+                        lp_qdisc->next = NULL;
```
The whole body of this `if` statement can be simplified using `__skb_dequeue_tail()`.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,309 @@
+	struct sk_buff *skb;
+	struct sk_buff_head *hpq = &q->qdiscs[q->highest_prio];
+
+	skb = skb_dequeue(hpq);
```
Style: (1) add the line `struct sk_buff *skb = skb_dequeue(hpq);` after line `struct sk_buff_head *hpq = &q->qdiscs[q->highest_prio];`, and (2) drop the current lines: `struct sk_buff *skb;` and `skb = skb_dequeue(hpq);`.

Can't we replace `skb_dequeue()` with `__skb_dequeue()`? Does any scheduler aquire lock(s) while it enqueues and dequeues socket buffers to its datastructure?

ndev2 replied:
Yes, will make the change. Since the api for schedulers automatically locks the scheduler for qdisc ops, if we'd need to change variables/auxillary datastructures outside of the qdisc  itself, then we'd have to accquire the existing lock appropriately. Accessing the qdisc is taken care off by existing kernel code, none of the current scheduler worries about that while enqueueing or dequeueing.

AltraMayor reviewed (changes requested):
**net/sched: add gkprio scheduler**

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,309 @@
+		 __qdisc_drop(to_drop, to_free);
+		qdisc_qstats_drop(sch);
+        }	
+	else goto drop;
```
Style: put `goto drop;` on its own line, that is, break the line above after the `else`.

AltraMayor reviewed (changes requested):
**net/sched: add gkprio scheduler**

ndev2 commented:
Hi everyone, I forgot to make a comment with the commit, but here I've changed the `sk_buff_head` functions we use and style changes for the code as suggested.

ndev2 commented:
Hi everyone, here is  the drafted commit message for this patch, do let me know what you think:

commit subject: net/sched: add gkprio scheduler 

  gkprio (Gatekeeper Priority Queue) is a defense against denial-of-service attacks.

This patch adds the Gatekeeper Priority Queue (GK) scheduler. GK enforces prioritised packet scheduling with priorities assigned according to - DSCP values (in IP packets) and/or user-specified values (for non IP packets). A global limit specified to GK is distributed to individual priority queues as required, at the same time enforcing global packet prioritisation, eliminating the need to limit individual queues independently.  When the queue is full, the lowest priority packet in the queue is dropped to make room for the ingress packet if it has higher priority; if the ingress packet has a priority lesser than or equal to the current lowest priority packet in the queue, it is dropped.
  

GK is especially useful for denial-of-service defense service: it prioritises packets in flows that demonstrate expected behaviour of legitimate users. The queue is flexible to allow any number of packets of any priority up to the global limit of the scheduler without risking resource overconsumption by a flood of low priority packets.

Signed-off-by: Nishanth Devarajan <ndev_2021@gmail.com>
Reviewed-by: Sachin Paryani <sachin.paryani@gmail.com>
Reviewed-by: Cody Doucette <doucette@bu.edu>
Reviewed-by: Michel Machado <michel@digirati.com.br>

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,300 @@
+			return -EINVAL;
+		else
+			q->max_limit = ctl->limit;
+
```
Let's guarantee that `q->max_limit` is greater than zero adding the following `if` statement here:
```
if (!q->max_limit)
	return -EINVAL;
```

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,300 @@
+	return NET_XMIT_SUCCESS;
+drop:
+	qdisc_drop(skb, sch, to_free);
+	return NET_XMIT_SUCCESS | __NET_XMIT_BYPASS;
```
1. What does the flag `__NET_XMIT_BYPASS` mean?

2. Why do we return `NET_XMIT_SUCCESS | __NET_XMIT_BYPASS` instead of `NET_XMIT_DROP` here? If we can/should return `NET_XMIT_DROP`, we can rewrite both lines above as `return qdisc_drop(skb, sch, to_free);`.

ndev2 replied:
Hi Michel, `__NET_XMIT_BYPASS` is a flag mask used by the kernel to indicate that the packet did not meet/have the requirements to be processed by the releant qdisc, i.e it bypassed the qdisc functionality processing, and was dropped. Other schedulers use it in the same way. Let me know if I should change this, thanks.

AltraMayor replied:
Thanks for the good explanation. We can keep these lines as they are now.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,300 @@
+		return NET_XMIT_SUCCESS;
+	}
+
+	/* If this packet is lowest priority, drop it. */
```
Small adjust of the comment: `/* If this packet has the lowest priority, drop it. */`

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,300 @@
+	lp_qdisc = &q->qdiscs[lp];
+	to_drop = __skb_dequeue_tail(lp_qdisc);
+
+        if (likely(to_drop)) {
```
Thanks to another comment in this review, we guarantee that `q->max_limit >= 1`, so `to_drop` cannot be `NULL` here. Instead of having the `if` statement above, we can (1) write `BUG_ON(!to_drop);`, and (2) drop lines `if (likely(to_drop)) {`, `}`, and `else goto drop;`.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,300 @@
+
+        if (likely(to_drop)) {
+		qdisc_qstats_backlog_dec(sch, to_drop);
+		qdisc_qstats_drop(sch);
```
Bug: you want to replace `qdisc_qstats_drop()` with `qdisc_drop()`. Otherwise, `to_drop` is memory leaked.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,300 @@
+
+	/* Update highest priority field. */
+	if (hpq->qlen == 0)
+		q->highest_prio = calc_new_high_prio(q);
```
BUG: consider the case that there was a single packet in the queue when `gkprio_dequeue()` was called. This means that `q->lowest_prio == q->highest_prio`, but we only update `q->highest_prio` here.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,300 @@
+		q->lowest_prio = calc_new_low_prio(q);
+
+	if (prio > q->highest_prio)
+		q->highest_prio = prio;
```
BUG: consider the case that there was a single packet in the queue when `gkprio_enqueue()` was called. This means that `q->lowest_prio == q->highest_prio`, but the `if` statement before the one above only updates `q->lowest_prio`, and the `if` statement above will only update `q->highest_prio` if the priority of the new packet is higher than before.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,300 @@
+	int prio;
+	
+	for (prio = q->highest_prio - 1; prio >= q->lowest_prio; prio--) {	
+		if (q->qdiscs[prio].qlen > 0)
```
Replace expression `q->qdiscs[prio].qlen > 0` with `!skb_queue_empty(&q->qdiscs[prio])` for a slight readbility improvement.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,300 @@
+	int prio;
+
+	for (prio = q->lowest_prio + 1; prio <= q->highest_prio; prio++) {
+		if (q->qdiscs[prio].qlen > 0)
```
Replace expression `q->qdiscs[prio].qlen > 0` with `!skb_queue_empty(&q->qdiscs[prio])` for a slight readbility improvement.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,300 @@
+	qdisc_qstats_backlog_inc(sch, skb);
+
+	/* Check to update highest and lowest priorities. */
+	if (lp_qdisc->qlen == 0)
```
Replace expression `lp_qdisc->qlen == 0` with `skb_queue_empty(lp_qdisc)` for a slight readbility improvement.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,300 @@
+	qdisc_bstats_update(sch, skb);
+
+	/* Update highest priority field. */
+	if (hpq->qlen == 0)
```
Replace expression `hpq->qlen == 0` with `skb_queue_empty(hpq)` for a slight readbility improvement.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,300 @@
+	if (prio <= lp)
+		return qdisc_drop(skb, sch, to_free);
+
+	/* Drop the packet at the tail of lowest priority qdisc. */
```
Small adjust of the comment (it adds a missing "the"): `/* Drop the packet at the tail of the lowest priority qdisc. */`

AltraMayor reviewed (changes requested):
**net/sched: add gkprio scheduler**

AltraMayor commented:
I suggest the following commit message:
```
net/sched: add gkprio scheduler

Gkprio (Gatekeeper Priority Queue) is a queueing discipline that
prioritizes IPv4 and IPv6 packets accordingly to their DSCP field.
Although Gkprio can be employed in any QoS scenario in which
a higher DSCP field means a higher priority packet,
Gkprio was conceived as a solution for denial-of-service defenses that
need to route packets with different priorities.

Signed-off-by: Nishanth Devarajan <ndev_2021@gmail.com>
Reviewed-by: Sachin Paryani <sachin.paryani@gmail.com>
Reviewed-by: Cody Doucette <doucette@bu.edu>
Reviewed-by: Michel Machado <michel@digirati.com.br>
```

ndev2 commented:
Hi Michel, I've made the suggested changes, hope its good. Will do, I will use the suggested commit message while sending out the patch. Thanks.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,306 @@
+	/* Update highest priority field. */
+	if (skb_queue_empty(hpq))
+		if (q->lowest_prio == q->highest_prio) {
+			q->highest_prio = 0;
```
Add `BUG_ON(sch->q.qlen);` between the two lines above. This assert helps us to catch bugs earlier, and even improve readability of the code since it makes clear that the queue must be empty.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,306 @@
+	/* Check to update highest and lowest priorities. */
+	if (skb_queue_empty(lp_qdisc))
+		if (q->lowest_prio == q->highest_prio)
+			q->lowest_prio = q->highest_prio = calc_new_low_prio(q);
```
1. Add `BUG_ON(sch->q.qlen);` immediatly before the line above. You'll need to add braces to the body of the `if` statement.

2. Replace `calc_new_low_prio(q);` with `prio`.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,306 @@
+	/* Drop the packet at the tail of the lowest priority qdisc. */
+	lp_qdisc = &q->qdiscs[lp];
+	to_drop = __skb_dequeue_tail(lp_qdisc);
+
```
Drop this blank line. This will make the comment above explain the whole block of code.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,306 @@
+		else
+			q->max_limit = ctl->limit;
+		if (!q->max_limit)
+			return -EINVAL;		
```
Place the two lines above just after the body of the `else` where it is right now. This will guarantee that if `opt` is `NULL` and `qdisc_dev(sch)->tx_queue_len` is zero, we will return `-EINVAL`.

I know that I've asked to place these lines here, but I didn't notice the issue above before. Sorry.

ndev2 replied:
Made the change. No worries.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,306 @@
+	.peek		=	qdisc_peek_dequeued,
+	.init		=	gkprio_init,
+	.reset		=	gkprio_reset,
+	.change		=	gkprio_init,
```
`gkprio_init()` updates the parameters *and* initializes the queues, whereas `.change` expects to only update the parameters.The problem can be easy addressed breaking up `gkprio_init()` into two functions.

AltraMayor reviewed (changes requested):
**net/sched: add gkprio scheduler//Fix bugs, style changes**

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,306 @@
+	sch->q.qlen = 0;
+
+	for (prio = 0; prio < GKPRIO_MAX_PRIORITY; prio++)
+		skb_queue_purge(&q->qdiscs[prio]);
```
Use `__skb_queue_purge()`.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,306 @@
+	int prio;
+
+	for (prio = 0; prio < GKPRIO_MAX_PRIORITY; prio++)
+		skb_queue_purge(&q->qdiscs[prio]);
```
Use `__skb_queue_purge()`.

AltraMayor reviewed (changes requested):
**net/sched: add gkprio scheduler**

ndev2 commented:
Hi Michel, made the suggested changes, hope they're good. Thanks.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,315 @@
+{
+	struct gkprio_sched_data *q = qdisc_priv(sch);
+
+	if (!opt) {
```
Notice that `gprio_change()` always receives `opt`, so this test really should go to `gkprio_init()`. The `else` part of the `if` statement that belongs to `gprio_change()`.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,315 @@
+		if (!q->max_limit)
+			return -EINVAL;
+		else
+			q->max_limit = ctl->limit;
```
Make this `else` clause part of the previous `if`. Otherwise `ctl->limit == 0` will make `q->max_limit` zero.

Notice that the change made is not what I suggested in my previous review, but, due my `max()` comment in this review, that no longer matters.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,315 @@
+	struct gkprio_sched_data *q = qdisc_priv(sch);
+
+	if (!opt) {
+		q->max_limit = qdisc_dev(sch)->tx_queue_len;
```
Replace `qdisc_dev(sch)->tx_queue_len` with `max(qdisc_dev(sch)->tx_queue_len, 1)`. This will guarantee that `q->max_limit` is not zero when `opt` is `NULL`.

AltraMayor reviewed (changes requested):
**net/sched: add gkprio scheduler**

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,315 @@
+			return -EINVAL;
+
+		if (ctl->limit == (typeof(ctl->limit)) - 1)
+			q->max_limit = qdisc_dev(sch)->tx_queue_len;
```
Replace `qdisc_dev(sch)->tx_queue_len` with `max(qdisc_dev(sch)->tx_queue_len, 1)`. This will match this case with the case in which `opt` is `NULL` (see previous round of comments).

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,315 @@
+		if (ctl->limit == (typeof(ctl->limit)) - 1)
+			q->max_limit = qdisc_dev(sch)->tx_queue_len;
+		else if (ctl->limit > qdisc_dev(sch)->tx_queue_len)
+			return -EINVAL;
```
Replace the test `ctl->limit > qdisc_dev(sch)->tx_queue_len` with `ctl->limit < 1 || ctl->limit > qdisc_dev(sch)->tx_queue_len`. This will simplify the code a bit (see next comment).

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,315 @@
+			return -EINVAL;
+
+		if (!q->max_limit)
+			return -EINVAL;
```
Thanks to the new test `ctl->limit < 1 || ctl->limit > qdisc_dev(sch)->tx_queue_len` in the previous comment, we can drop this `if` statement. BUT preserve the `else` clause attaching it to the previous `if` statement.

AltraMayor reviewed (changes requested):
**net/sched: add gkprio scheduler**

ndev2 commented:
Hi Michel, I've made the suggested changes, hope they're good. Thanks :)

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,313 @@
+		q->noip_dfltp = 0;
+		return 0;
+	}
+	else return gkprio_change(sch, opt);
```
Style: drop the keyword `else`, it is redundant due to the `return` in the then clause of the `if` statement.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,313 @@
+	if (ctl->limit == (typeof(ctl->limit)) - 1)
+		q->max_limit = max(qdisc_dev(sch)->tx_queue_len, 1);
+	else if (ctl->limit < qdisc_dev(sch)->tx_queue_len ||
+		ctl->limit >qdisc_dev(sch)->tx_queue_len)
```
If you go back to my comment in the previous review (https://github.com/ndev2/net-next/pull/2#discussion_r179927036), you'll notice that the correct expression got lost: `ctl->limit < 1 || ctl->limit > qdisc_dev(sch)->tx_queue_len`

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,313 @@
+	struct gkprio_sched_data *q = qdisc_priv(sch);
+	int prio;
+
+	/* Initialise 64 queues, one for each possible priority. */
```
Replace `64` with `all`.

AltraMayor commented on include/uapi/linux/pkt_sched.h:
```diff
@@ -124,6 +124,17 @@ struct tc_fifo_qopt {
+
+struct tc_gkprio_qopt {
+	__u32	limit; 	    	/* Queue length in packets. */
+	__u16	noip_dfltp; 	/* Default priority for Non-IP packets. */
```
Make the `N` in `Non-IP` lowercase: `non-IP`.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,313 @@
+ * This qdisc schedules a packet according to the value (0-63) of its DSCP
+ * (IPv4) or DS (IPv6) field, where a higher value places the packet closer
+ * to the exit of the queue. Non-IP packets are assigned a default priority
+ * specified to Gatekeeper; if none is specified, default priority is set
```
Replace `Gatekeeper` with `GKPRIO` just before the semicolon.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,313 @@
+ * of individual priorities.
+ *
+ * This is especially useful for a denial-of-service defense service,
+ * Gatekeeper, which prioritizes packets in flows that demonstrate expected
```
add `like` just before `Gatekeeper`.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,313 @@
+ * queue length limit can be enforced while not restricting the queue lengths
+ * of individual priorities.
+ *
+ * This is especially useful for a denial-of-service defense service,
```
Replace `service,` with `system`; notice that I'm suggesting to drop the comma as well.

AltraMayor commented on net/sched/sch_gkprio.c:
```diff
@@ -0,0 +1,313 @@
+			q->highest_prio = 0;
+			q->lowest_prio = GKPRIO_MAX_PRIORITY - 1;
+		}
+		else
```
Style: combine both lines into a single line: `} else`.

AltraMayor reviewed (changes requested):
**net/sched: add gkprio scheduler**

Congratulations, this is my last review. After addressing the comments in this last review, you should do the following:

1. Check style with `scripts/checkpatch.pl`. See section 4 of the document "Submitting patches: the essential guide to getting your code into the kernel" that I referred you before.

2. Test the code. This is time-consuming, but the more you test, the less the chance that there's a bug in the code.

3. Follow the document "Submitting patches: the essential guide to getting your code into the kernel" **to the letter**.

Once you've done the items above, you are all set to push your patch upstream. Good luck!

ndev2 commented:
Hi Michel, here is the `scripts/scheckpatch.pl` checks:

```
CHECK: Avoid crashing the kernel - try using WARN_ON & recovery code rather than BUG() or BUG_ON()
#152: FILE: net/sched/sch_gkprio.c:152:
+	BUG_ON(!to_drop);

CHECK: Avoid crashing the kernel - try using WARN_ON & recovery code rather than BUG() or BUG_ON()
#162: FILE: net/sched/sch_gkprio.c:162:
+			BUG_ON(sch->q.qlen);

CHECK: Avoid crashing the kernel - try using WARN_ON & recovery code rather than BUG() or BUG_ON()
#195: FILE: net/sched/sch_gkprio.c:195:
+			BUG_ON(sch->q.qen);

WARNING: else is not generally useful after a break or return
#236: FILE: net/sched/sch_gkprio.c:236:
+		return -EINVAL;
+	else

WARNING: else is not generally useful after a break or return
#243: FILE: net/sched/sch_gkprio.c:243:
+		return -EINVAL;
+	else

total: 0 errors, 2 warnings, 3 checks, 315 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

net/sched/sch_gkprio.c has style problems, please review.

NOTE: If any of the errors are false positives, please report
      them to the maintainer, see CHECKPATCH in MAINTAINERS.
```
I've changed a few lines of code to remove style issues pointed out by the script. But the above checks and warnings still remain.  Hope the code is good,  let me know if  we should change something. Will test the code,  and get back to you with the results I find. Thanks.

AltraMayor commented:
The remaining feedback of `scripts/scheckpatch.pl` can be safely ignored because there's no recovery code for our `BUG_ON()`s and the pointed  `else` clauses are needed.

You are all set to start the review of your code upstream.
